data/pools/transactionPool.go

```diff
@@ -179,6 +179,44 @@ func (pool *TransactionPool) Remember(t transactions.SignedTxn) error {
 	defer pool.mu.Unlock()
 
 	err := pool.test(t)
+	// Check for a race condition when a new block was added
+	// to our ledger, but hasn't propagated to OnNewBlock()
+	// yet.
+	switch errt := err.(type) {
+	case transactions.TxnDeadError:
+		if errt.Round >= errt.FirstValid {
+			// This is not an error due to the round being too old.
+			break
+		}
+
+		if pool.pendingBlockEvaluator == nil {
+			// We don't have a block evaluator at all; something
+			// else went wrong, and this isn't the place to fix
+			// it.  (The next OnNewBlock callback might recover.)
+			break
+		}
+
+		latest := pool.ledger.Latest()
+		if pool.pendingBlockEvaluator.Round() > latest {
+			// Our block evaluator is already on a block after
+			// the latest block in the ledger.  No race to fix.
+			break
+		}
+
+		blk, err0 := pool.ledger.Block(latest)
```

Comment:
I'ma bit hesitant about this section; re-doing what should be done by the OnNewBlock() doesn't seems to be correct. If we have an issue there, it should be fixed there. Instead, we should use synchronization primitives ( somehow ) to wait until the OnNewBlock() arrives and making the correct adjustment. ( with a timeout in case of a server failure ).

Reply:
Indeed, another option would be to wait for the right `OnNewBlock()`.  I was thinking it's a bit more prone to deadlock if the `OnNewBlock()` never arrives for some reason, and the lack of a timed `Cond.Wait()` in Go pushed me over the edge to the code in this PR..

Let me sketch out what this condvar-based approach would look like and see if it's cleaner.

Reply:
thanks!